### PR TITLE
build: validate the source directory passed to ProjectBuilder

### PR DESCRIPTION
**Pull request by pfmoore**

### CODE REVIEW

**Comment by pfmoore** [comment]
Fixes #259.

Initial version, for some reason the `tox -e fix` command doesn't work on my PC (it doesn't recognise my copy of Visual Studio when building the pre-commit environments) so I can't run the linting locally. I'll let CI do it for me...

**Comment by pfmoore** [comment]
I don't know why the tests are failing, they work locally.

File: src/build/__init__.py [hunk 1]
```diff
@@ -59,6 +59,16 @@ class TypoWarning(Warning):
     """
 
 
+def _validate_source_directory(srcdir):
+    # type: (str) -> None
+    if not os.path.isdir(srcdir):
+        raise BuildException('Source {} is not a directory'.format(srcdir))
+    pyproject_toml = os.path.join(srcdir, "pyproject.toml")
+    setup_py = os.path.join(srcdir, "setup.py")
```
Comment (FFY00) [inline src/build/__init__.py:67]: You are not running pre-commit. Please use single quotes here :stuck_out_tongue:

File: src/build/__init__.py [hunk 2]
```diff
@@ -59,6 +59,16 @@ class TypoWarning(Warning):
...
+    if not (os.path.exists(pyproject_toml) or os.path.exists(setup_py)):
```
Comment (FFY00) [inline src/build/__init__.py:68]: nitpick: I find it easier to read in the expanded form using De Morgan's laws. I do not need to keep in mind that a logical block is open and mentally negate the statement.

**Review by FFY00 — APPROVED** [review]
I think checking for either `pyproject.toml` or `setup.py` is *fine*, even though I don't like that people with a `setup.cfg` will continue with the current behavior, making things slightly inconsistent, so I am not gonna block this.

File: src/build/__init__.py (hunk 1 above)
Comment (pfmoore) [reply src/build/__init__.py]: Thanks. Is the right way to run pre-commit to use the `tox -e fix` command? As I said, that's failing for me :-( Is there a contributing guide that explains how to set up for development, run tests, etc? I wasn't able to find one.

I tried running it by hand, which seems to have worked. So I've pushed a fix.

File: src/build/__init__.py (hunk 2 above)
Comment (pfmoore) [reply src/build/__init__.py]: Also fixed. Thanks!

File: src/build/__init__.py (hunk 1 above)
Comment (FFY00) [reply src/build/__init__.py]: It should, I don't know exactly what went wrong in your case :confused:. Perhaps @gaborbernat might have more insight?

Comment (FFY00) [reply src/build/__init__.py]: We should have a contributor guide, but right now I'm ETIME.

File: src/build/__init__.py (hunk 2 above)
Comment (FFY00) [reply src/build/__init__.py]: Thanks :blush:

File: src/build/__init__.py (hunk 1 above)
Comment (pfmoore) [reply src/build/__init__.py]: No worries, I completely understand.

**Comment by FFY00** [comment]
You seem to have found an issue with our integration tests :sweat_smile:. We test building a few popular projects in the CI as a sanity check and it looks like we are invoking `build` in the directory where the github tarballs are extracted, not the folder that is inside them and contains the project.

I will open a separate PR, you can then rebase on top of it.

File: src/build/__init__.py (hunk 1 above)
Comment (gaborbernat) [reply src/build/__init__.py]: > Thanks. Is the right way to run pre-commit to use the `tox -e fix` command? As I said, that's failing for me

Can you provide more exact details for this?

**Comment by FFY00** [comment]
Btw, as the integration tests take a long time and are only included as a sanity check, you need to pass `--only-integration` or `--run-integration` to pytest. You can do this in tox as `tox -e py39 -- --only-integration`.

My PR will introduce a test dependency so you will need to recreate the tox environment, you can do so with `tox -r -e py39`

Sorry for the trouble!

**Comment by gaborbernat** [comment]
> I don't know why the tests are failing, they work locally.

This is why I'm not the biggest fan of not running integration tests locally by default 😆 People think if it passed locally they are done.

**Review by gaborbernat — CHANGES_REQUESTED** [review]
The logic looks ok, but we need tests for these cases.

File: src/build/__init__.py (hunk 1 above)
Comment (pfmoore) [reply src/build/__init__.py]: I'll try to later, but basically it said I didn't have Visual C version 14 installed (I have VS 2019) when building the virtualenvs. But when I ran pre-commit outside of tox, it worked, and now `tox -e fix` works.

**Comment by pfmoore** [comment]
Yep, once the current tests pass I'll add tests for the new functionality.

**Comment by FFY00** [comment]
The tests should now be fixed now, @pfmoore can you rebase your branch on main?

**Comment by pfmoore** [comment]
Thanks for the quick fix, rebase done.

**Comment by FFY00** [comment]
Thanks, when you come back to add the tests, could you also add an entry in the changelog?

**Comment by pfmoore** [comment]
Done.

By the way, the symlink from `docs/changelog.rst` to `CHANGELOG.rst` doesn't get extracted properly on Windows, and messes up the git repo. Twice, I nearly uploaded a commit which changed that file to a plain text file containing the text `../CHANGELOG.rst` 🙁 

For portability, it might be better not to include symlinks in the repository...

File: src/build/__init__.py (hunk 1 above)
Comment (gaborbernat) [reply src/build/__init__.py]: Weird. Was it trying to build some c-extension package (such as black?)

**Comment by FFY00** [comment]
Uh, I thought symlinks were fixed on Windows, I guess that's not the case.

File: CHANGELOG.rst [hunk 3]
```diff
@@ -3,6 +3,15 @@ Changelog
 +++++++++
 
 
+0.3.2 (XX-XX-2021)
+==================
+
+- Validate that the supplied source directory is valid (`PR #260`_)
+
+.. _PR 260: https://github.com/pypa/build/pull/260
```
Comment (FFY00) [inline CHANGELOG.rst:11]: Missing a `#` here, `PR #260`.

Could you also add a "Fixes #259"?

**Review by FFY00 — APPROVED** [review]
Other than this, it should be good! Thanks :blush:

File: src/build/__init__.py (hunk 1 above)
Comment (pfmoore) [reply src/build/__init__.py]: Yes. I can't recall which one, I'll let you know once I manage to reproduce the issue.

**Comment by pfmoore** [comment]
I do have symlinks enabled, and they do work. I suspect that the problem is that git might not handle them. I'm using a very vanilla git setup.

**Comment by pfmoore** [comment]
@gaborbernat I've raised #262 with details of the pre-commit failure.

@FFY00 I think I've done everything needed now. I've no idea why the coverage has gone down (or what I need to do to fix that). The coverage report seems to be flaggking lines that I haven't touched as uncovered 🙁 

Let me know what (if anything) more I need to do.

**Comment by FFY00** [comment]
Thanks! The coverage is a bit fiddly, it is fine now.

**Comment by pfmoore** [comment]
Thanks @FFY00 and @gaborbernat for the help getting this PR through 🙂 Much appreciated.

File: tests/test_projectbuilder.py [hunk 4]
```diff
@@ -136,6 +136,18 @@ def test_check_dependency(monkeypatch, requirement_string, expected):
     assert next(build.check_dependency(requirement_string), None) == expected
 
 
+def test_bad_project(test_no_project_path):
+    # Passing a nonexistent project directory
+    with pytest.raises(build.BuildException):
```
Comment (gaborbernat) [inline tests/test_projectbuilder.py:141]: Can you split each test into its own test please? 👌

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Otherwise looks good.

File: tests/test_projectbuilder.py (hunk 4 above)
Comment (FFY00) [reply tests/test_projectbuilder.py:141]: The code already went in. This is not a big deal though.

**Comment by henryiii** [comment]
Sorry I missed this. It's also valid to have setup.cfg but no setup.py or pyproject.toml.

**Comment by henryiii** [comment]
See: https://github.com/joerick/cibuildwheel/blob/96aa38b56c241773f612ee702d638eb76c80576b/cibuildwheel/__main__.py#L164-L169

**Comment by layday** [comment]
It's not for legacy projects - how are you going to be able to invoke `sdist` or `bdist_wheel`?  If it works in pip it must be because it calls `setup()` on the fly.

**Comment by pfmoore** [comment]
If there's no `pyproject.toml` then (in certain circumstances - it's complicated with all sorts of compatibility cruft) pip will use the `setuptools.build_meta:__legacy__` backend, as suggested [in PEP 517](https://www.python.org/dev/peps/pep-0517/#source-trees).

**Comment by henryiii** [comment]
Not having a pyproject.toml is not something I do, but setup.cfg only projects are valid, and currently work in `build`:

```bahs
cookiecutter gh:scikit-hep/cookie
rm pyproject.toml
rm setup.py
pipx run build
```

This correctly produces a wheel and SDist, just like pip would. (Technically, I'm not sure how you would make an SDist without build, actually). Current master would now be broken. We suggest setup.cfg only projects on https://packaging.python.org/tutorials/packaging-projects/ (with a pyproject.toml too, of course, but technically it can be omitted and pip and current build work).

**Comment by henryiii** [comment]
Quick followup, pip checks this:

```bash
python3 -m pip wheel .
ERROR: Directory '.' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.
```

So being just as picky in build is perfectly fine. Though it is a removal of something that used to work.

**Comment by henryiii** [comment]
As of https://github.com/pypa/pip/pull/9945, we've now swapped with Pip. Pip 21.1.2 _does_ support _just_ setup.cfg now, and now we don't. 🤣

```console
$ pip wheel .
ERROR: Directory '.' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.
WARNING: You are using pip version 21.1.1; however, version 21.1.2 is available.
You should consider upgrading via the '/Users/henryschreiner/tmp/quickcheck/venv/bin/python3.9 -m pip install --upgrade pip' command.
$ pip install --upgrade pip
$ pip wheel .
Processing /Users/henryschreiner/tmp/quickcheck
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
Collecting numpy>=1.13.3
  File was already downloaded /Users/henryschreiner/tmp/quickcheck/numpy-1.20.3-cp39-cp39-macosx_10_9_x86_64.whl
Building wheels for collected packages: quickcheck
  Building wheel for quickcheck (setup.py) ... done
  Created wheel for quickcheck: filename=quickcheck-0.0.0-py3-none-any.whl size=3372 sha256=57acfb84109eedde21404ffdea6ef0e87d6ef7663498e3220bb0c4a72b066d35
  Stored in directory: /private/var/folders/_8/xtbws09n017fbzdx9dmgnyyr0000gn/T/pip-ephem-wheel-cache-62ksr1uu/wheels/9a/34/3d/e8d9298f424ae357bac078a56990fce39937c7ed9018594f47
Successfully built quickcheck
```

**Pull request merged**